docs/guides/typescript.md

```diff
@@ -233,3 +203,258 @@ type State =
+      }
+    },
+    state2: {
+      entry: [
+        (context, event) => {
+          // This will error at .flag
+          console.log(event.flag);
+        }
+      ]
+    }
+  }
+});
+```
+
+Here, we don't know what event led to the `entry` action on `state2`. The only way to fix this is to do a similar trick to above:
+
+```ts
+entry: [
+  (context, event) => {
+    if (event.type !== 'EVENT_WITH_FLAG') return;
+    // No more error at .flag!
+    console.log(event.flag);
+  }
+];
+```
+
+### `onDone`/`onError` events in machine options
+
+The result of promise-based services is quite hard to type safely in XState. For instance, a machine like this:
+
+```ts
+interface Data {
+  flag: boolean;
+}
+
+interface Context {}
+
+type Event = {
+  // Added here in order to bring out the TS errors
+  type: 'UNUSED_EVENT';
+};
+
+createMachine<Context, Event>(
+  {
+    invoke: {
+      src: async () => {
+        const data: Data = {
+          flag: true
+        };
+        return data;
+      },
+      onDone: {
+        actions: 'consoleLogData'
+      },
+      onError: {
+        actions: 'consoleLogError'
+      }
+    }
+  },
+  {
+    actions: {
+      consoleLogData: (context, event) => {
+        // Error on this line - data does not exist!
+        console.log(event.data.flag);
+      },
+      consoleLogError: (context, event) => {
+        // Error on this line - data does not exist!
+        console.log(event.data);
+      }
+    }
+  }
+);
+```
+
+Frustratingly, the best way to fix this is to cast the `event` to `any` and reassign it based on what we know it to be:
+
+```ts
+import { DoneInvokeEvent, ErrorPlatformEvent } from 'xstate'
+
+actions: {
+  consoleLogData: (context, _event: any) => {
+    const event: DoneInvokeEvent<Data> = _event;
+    console.log(event.data.flag);
+  },
+  consoleLogError: (context, _event: any) => {
+    const event: ErrorPlatformEvent = _event;
+    // Event.data is usually of type `Error`
+    console.log(event.data.message);
+  }
+}
+```
+
+### Assign action behaving strangely
+
+When run in `strict: true` mode, assign actions can sometimes behave very strangely.
+
+```ts
+interface Context {
+  something: boolean;
+}
+
+createMachine<Context>({
+  context: {
+    something: true
+  },
+  entry: [
+    // Type 'AssignAction<{ something: false; }, AnyEventObject>' is not assignable to type 'string'.
+    assign(() => {
+      return {
+        something: false
+      };
+    }),
+    // Type 'AssignAction<{ something: false; }, AnyEventObject>' is not assignable to type 'string'.
+    assign({
+      something: false
+    }),
+    // Type 'AssignAction<{ something: false; }, AnyEventObject>' is not assignable to type 'string'.
+    assign({
+      something: () => false
+    })
+  ]
+});
+```
+
+It might appear that nothing you try works - all syntaxes are buggy. The fix is very strange, but works consistently. Add an unused `context` argument to the first argument of your assigner function.
```

Comment:
FWIW I've seen this forever down in `services`. Same behaviour, same fix.

Reply:
Hmmm, weird!

Reply:
Could well have been me not knowing what I was doing, been a while since I’ve seen it and can’t replicate in a minimal CSB.